synapse/handlers/room.py

```diff
@@ -1217,6 +1218,114 @@ async def filter_evts(events: List[EventBase]) -> List[EventBase]:
         return results
 
 
+class TimestampLookupHandler:
+    def __init__(self, hs: "HomeServer"):
+        self.server_name = hs.hostname
+        self.store = hs.get_datastore()
+        self.state_handler = hs.get_state_handler()
+        self.transport_layer = hs.get_federation_transport_client()
+
+    async def get_event_for_timestamp(
+        self,
+        requester: Requester,
+        room_id: str,
+        timestamp: int,
+        direction: str,
+    ) -> Optional[str]:
+        """Find the closest event to the given timestamp in the given direction.
+        If we can't find an event locally or the event we have locally is next to a gap,
+        it will ask other federated homeservers for an event.
+
+        Args:
+            requester: The user making the request according to the access token
+            room_id: Room to fetch the event from
+            timestamp: The point in time we should navigate from in the given
+                direction to find the closest event.
+            direction: ["f"|"b"] to indicate whether we should navigate forward
+                or backward from the given timestamp to find the closest event.
+
+        Returns:
+            `event_id` closest to the given timestamp in the given direction
+
+        Raises:
+            SynapseError if unable to find any event locally in the given direction
+        """
+
+        local_event_id = await self.store.get_event_for_timestamp(
+            room_id, timestamp, direction
+        )
+
+        logger.debug(
+            "get_event_for_timestamp: locally, we found event_id=%s closest to timestamp=%s",
+            local_event_id,
+            timestamp,
+        )
+
+        # If we found an extremity, we should probably ask another homeserver
+        # first about more history in between
+        is_extremity = False
+        if local_event_id:
+            is_extremity = await self.store.check_if_event_is_extremity(
+                room_id, local_event_id
+            )
+        if not local_event_id or is_extremity:
+            logger.debug(
+                "get_event_for_timestamp: locally, we found event_id=%s closest to timestamp=%s which is an extremity so we're asking other homeservers first",
+                local_event_id,
+                timestamp,
+            )
+
+            # Find other homeservers from the given state in the room
+            curr_state = await self.state_handler.get_current_state(room_id)
+            curr_domains = get_domains_from_state(curr_state)
+            likely_domains = [
+                domain for domain, depth in curr_domains if domain != self.server_name
+            ]
+
+            # Loop through each homeserver candidate until we get a succesful response
+            for domain in likely_domains:
+                try:
+                    remote_response = await self.transport_layer.timestamp_to_event(
+                        domain, room_id, timestamp, direction
+                    )
```

Comment:
If no homeserver has an answer for `timestamp_to_event`, they'll spam each other with queries recursively.

eg. if a client were to ask for the closest event before January 1st 1970 or after 2030

Reply:
It shouldn't do it recursively. Only the client endpoint goes and tries to ask other federated homeservers.

But the federation endpoint acts differently. When a homeserver asks via the federation endpoint, it only tries to get an event from it's own local database.

This means that someone making a `/client/timestamp_to_event` request will only result in a single `/federation/timestamp_to_event` request going out to each participating homeserver.

---

As a thought, for a hard-limit we could check that the timestamp comes after the `m.room.create` event. But with [MSC2716](https://github.com/matrix-org/matrix-doc/pull/2716) coming around the corner, events will have timestamps before the `m.room.create` event so that isn't a viable option.

Reply:
My mistake, I misread the code. Please ignore my comment.